cpp/include/cudf/io/datasource.hpp

```diff
@@ -0,0 +1,182 @@
+
+namespace cudf {
+//! IO interfaces
+namespace io {
+
+/**
+ * @brief Interface class for providing input data to the readers.
+ */
+class datasource {
+ public:
+  /**
+   * @brief Interface class for buffers that the datasource returns to the caller.
+   *
+   * Provides a basic interface to return the data address and size.
+   */
+  class buffer {
+   public:
+    /**
+     * @brief Returns the buffer size in bytes.
+     */
+    virtual size_t size() const = 0;
+
+    /**
+     * @brief Returns the address of the data in the buffer.
+     */
+    virtual const uint8_t* data() const = 0;
+
+    /**
+     * @brief Base class destructor
+     */
+    virtual ~buffer() {}
+  };
+
+  /**
+   * @brief Creates a source from a file path.
+   *
+   * @param[in] filepath Path to the file to use
+   * @param[in] offset Bytes from the start of the file (the default is zero)
+   * @param[in] size Bytes from the offset; use zero for entire file (the default is zero)
+   */
+  static std::unique_ptr<datasource> create(const std::string& filepath,
+                                            size_t offset = 0,
+                                            size_t size   = 0);
+
+  /**
+   * @brief Creates a source from a memory buffer.
+   *
+   * @param[in] data Pointer to the input data host buffer
+   * @param[in] size Size of the buffer in bytes
+   */
+  static std::unique_ptr<datasource> create(const char* data, size_t size);
+
+  /**
+   * @brief Creates a source from a from an Arrow file.
+   *
+   * @param[in] arrow_file RandomAccessFile to which the API calls are forwarded
+   */
+  static std::unique_ptr<datasource> create(
+    std::shared_ptr<arrow::io::RandomAccessFile> arrow_file);
+
+  /**
+   * @brief Creates a source from an user implemented datasource object.
+   *
+   * @param[in] source Non-owning pointer to the datasource object
+   */
+  static std::unique_ptr<datasource> create(datasource* source);
+
+  /**
+   * @brief Base class destructor
+   */
+  virtual ~datasource(){};
+
+  /**
+   * @brief Returns a buffer with a subset of data from the source.
+   *
+   * @param[in] offset Bytes from the start
+   * @param[in] size Bytes to read
+   *
+   * @return The data buffer
+   */
+  virtual std::unique_ptr<datasource::buffer> host_read(size_t offset, size_t size) = 0;
+
+  /**
+   * @brief Reads a selected range into a preallocated buffer.
+   *
+   * @param[in] offset Bytes from the start
+   * @param[in] size Bytes to read
+   * @param[in] dst Address of the existing host memory
+   *
+   * @return The number of bytes read (can be smaller than size)
+   */
+  virtual size_t host_read(size_t offset, size_t size, uint8_t* dst) = 0;
+
+  /**
+   * @brief Whether or not this source supports reading directly into device memory.
+   *
+   * If this function returns true, the datasource will receive calls to device_read() instead of
+   * host_read() when possible. However, it is still possible to receive host_read() calls as well.
```

Comment:
> However, it is still possible to receive host_read() calls as well.

Under what circumstances? Do they need to be documented?

Reply:
Any data that requires CPU processing will always be read using `host_read`. This would typically be the metadata. I'll expand the comment to make this clearer, but I'm not sure how much detail is needed here.

Reply:
modified the comment, hopefully it's clearer now